primer/src/Primer/Typecheck.hs

```diff
@@ -814,6 +814,10 @@ check t = \case
             Hole{} -> default_ -- Don't let the recursive call mint a hole.
             e'' -> pure e''
       (Hole _ (Ann _ _ ty), SmartHoles)
+        -- TODO/REVIEW: when testing editable-forall-kinds, I noticed that we would not elide the hole in
+        --   foo : ∀a:*.? ; foo = {? ? : ∀a:*.? ?}
+        -- but this is odd -- the annotations are identical (so not acting as a typechange)!
+        -- (also, note that we keep hole-wrapped-holey-anns around because of the below even when we would accept the non-wrapped version (indeed, have an action that will do that!)
```

Comment:
This is odd. DO you have any inkling as to why it occurs?

Reply:
Yes. This is actually #7. The current design is
- When synthesising "never elide holes", i.e. do not automatically do `{? e ?} ~> e` when we have no information about the context. NB: if we did elide, it could change the synthesised type and force new holes to appear elsewhere in the program.
- When checking `T ∋ {? e ?}` we attempt to elide the hole if that would be type-correct by the transformations
  - if checking nested holes `T ∋ {? {? e ?} ?}`, drop one layer and recurse `T ∋ {? e ?}`
  - if have an annotation in a hole (this is the case referred to in this comment) `T ∋ {? s : S ?}`, we
    - drop a hole-and-trivial-annotation, if it is still type-correct `{? s : ? ?}` changes to `s`, if we have `T ∋ s`
    - never touch holes-and-incomplete-annotations (this is our current example), i.e. if there is a type hole in the annotation `S`, as (quoting Typecheck.hs) 
      > Don't want to, e.g., remove {? λx.x : ? ?} to get λx.x : ? Since holey annotations behave like non-empty holes, we will not elide non-empty holes if they have a holey annotation. (This is needed for idempotency, since we return non-empty holes with holey-annotated contents in the case a construction cannot typecheck, e.g. Bool ∋ λx.t returns {? λx.t : ? ?}
    - otherwise fall through to the "have a hole, and don't care about the contents" case
  - if none of the above are true, we have a hole and don't care about the contents, in which case we elide the hole leaving `e` as long as that is type-correct

(NB: above, where I say "elide...as long as that is type correct" and similar, by "type correct" I actually mean "smart holes accepts the term and does not create a hole _at top level_", i.e. it could result in "pushing down the hole")

Reply:
Can you add a comment in the code here with a link to the parent post, and also a link to #7, please?

Reply:
> Can you add a comment in the code here with a link to the parent post, and also a link to #7, please?

Will do, but in a separate PR since it is orthogonal to the current changes.